### PR TITLE
fix(subscription): allow updating subscription tags and display name with AzApi

### PR DESCRIPTION
**Pull request by luke-taylor**
# Overview/summary

Add the ability to update tags and display name if using azapi to vend a subscription.

## Testing evidence

Please provide testing evidence to show that your Pull Request works/fixes as described and documented above.

image

## As part of this pull request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/terraform-azurerm-lz-vending/pulls)
- [x] Associated it with relevant [issues](https://github.com/Azure/terraform-azurerm-lz-vending/issues), for tracking and closure.
- [x] Run and `make fmt` & `make docs` to format your code and update documentation.
- [x] Created unit and deployment tests and provided evidence.
- [x] Updated relevant and associated documentation.

### CODE REVIEW

File: modules/subscription/main.tf [hunk 1]
```diff
@@ -35,4 +35,34 @@ resource "azapi_resource" "subscription" {
...
+}
+
+resource "azapi_update_resource" "subscription_tags" {
+  count = var.subscription_alias_enabled && var.subscription_use_azapi ? 1 : 0
+
+  type        = "Microsoft.Resources/tags@2022-09-01"
+  resource_id = "/subscriptions/${jsondecode(azapi_resource.subscription[0].output).properties.subscriptionId}/providers/Microsoft.Resources/tags/default"
+  body = jsonencode({
+    properties = {
+      tags = var.subscription_tags
```
Comment (jaredfholgate) [inline modules/subscription/main.tf:52]: Tags are still specified in the alias resource. Are they still needed there?

File: modules/subscription/main.tf [hunk 2]
```diff
@@ -35,4 +35,34 @@ resource "azapi_resource" "subscription" {
...
+  count = var.subscription_alias_enabled && var.subscription_use_azapi ? 1 : 0
+
+  type        = "Microsoft.Resources/subscriptions@2021-10-01"
+  resource_id = "/subscriptions/${jsondecode(azapi_resource.subscription[0].output).properties.subscriptionId}"
+  method      = "POST"
+  action      = "providers/Microsoft.Subscription/rename"
+  body = jsonencode({
+    subscriptionName = var.subscription_display_name
+  })
+  response_export_values = ["*"]
```
Comment (jaredfholgate) [inline modules/subscription/main.tf:67]: Is this response being used?

File: modules/subscription/main.tf (hunk 1 above)
Comment (luke-taylor) [reply modules/subscription/main.tf:52]: Could cause a breaking, if I removed the arg no?

File: modules/subscription/main.tf (hunk 2 above)
Comment (luke-taylor) [reply modules/subscription/main.tf:67]: Nope, can be dropped

File: modules/subscription/main.tf [hunk 3]
```diff
@@ -35,4 +35,33 @@ resource "azapi_resource" "subscription" {
...
+  type        = "Microsoft.Resources/tags@2022-09-01"
+  resource_id = "/subscriptions/${jsondecode(azapi_resource.subscription[0].output).properties.subscriptionId}/providers/Microsoft.Resources/tags/default"
+  body = jsonencode({
+    properties = {
+      tags = var.subscription_tags
+    }
+  })
+}
+
+resource "azapi_resource_action" "subscription_rename" {
```
Comment (matt-FFFFFF) [inline modules/subscription/main.tf:57]: On initial sub creation, is there an unnecessary POST request to rename the sub, even though the name hasn't changed?

**Review by matt-FFFFFF — CHANGES_REQUESTED** [review]
One question

File: modules/subscription/main.tf (hunk 3 above)
Comment (luke-taylor) [reply modules/subscription/main.tf:57]: Yes there will be, similarly an unnecessary updating of the tags.

File: modules/subscription/main.tf (hunk 1 above)
Comment (jaredfholgate) [reply modules/subscription/main.tf:52]: It is possible I am not understanding how AzAPI works. Does the alias create them and then the PUT is only used if there is an update? What happens to the alias resource if someone updates the tags? Won't it try to update that resource too? Same for subscription name?

Comment (luke-taylor) [reply modules/subscription/main.tf:52]: Creating an alias is a write-once scenario as I have been told, so updates do nothing practically after the first apply, but do cause the terraform to think there's an update causing a bit of noise.

Comment (jaredfholgate) [reply modules/subscription/main.tf:52]: Right, got it. I see now. Ok, LGTM then.

**Review by jaredfholgate — DISMISSED** [review]
LGTM

**Review by matt-FFFFFF — APPROVED** [review]

**Comment by matt-FFFFFF** [comment]
LGTM

**Comment by jczerniak** [comment]
Hi, can we make it depend on the subscription_display_name variable and not on the alias_subscription_enabled variable. Without an alias we can still change subscription_display_name. We are not able to create an alias in every situation (subscriptions provided by a CSP partner), but we can always change the display name. 
Best Regards
Jacek

**Comment by matt-FFFFFF** [comment]
> Hi, can we make it depend on the subscription_display_name variable and not on the alias_subscription_enabled variable. Without an alias we can still change subscription_display_name. We are not able to create an alias in every situation (subscriptions provided by a CSP partner), but we can always change the display name. Best Regards Jacek

Hi @jczerniak 

Can you confirm the use case here?

Is it that you want to be able to update an existing subscription's tags and display name using this module?

**Comment by matt-FFFFFF** [comment]
Thanks for confirming @jczerniak 

We are discussing options here and want to propose the following condition. We see a risk in overwriting existing subscription tags and display name as defaults are empty string. Hence proposing a new input bool (`var.subscription_update_existing`) to control this.

Thoughts?

```
count = (var.subscription_alias_enabled && var.subscription_use_azapi) || (var.subscription_id != "" && var.subscription_update_existing)
```

**Comment by luke-taylor** [comment]
Latest Test after update 

image

**Review by matt-FFFFFF — APPROVED** [review]

**Comment by matt-FFFFFF** [comment]
I'll fix the test pipeline tomorrow. Sorry Luke. Great work here

**Comment by matt-FFFFFF** [comment]
@luke-taylor this is now ready. I have updated the tests.

Can you take a look and see if you are happy?

**Review by matt-FFFFFF — APPROVED** [review]

**Comment by luke-taylor** [comment]
> @luke-taylor this is now ready. I have updated the tests.
> 
> Can you take a look and see if you are happy?

Good catches on the test updates, LGTM.

**Pull request merged**